---
title: "TSEA, MSEA, etc"
author: "Fernando Lucas Ruiz (fernando.lucas@um.es)"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    highlight: kate
    number_sections: true
    theme: spacelab
    toc: true
    toc_float: true
    code_folding: "hide"
  pdf_document:
    toc: true
subtitle: Cirugía digestiva, endocrina y trasplante de órganos abdominales (IMIB)
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T, warning = T, message = F)
```

```{r}
source("../Librerias.R")
```

Cogemos los generos del maaslin para utilizarlo luego en el filtrado

```{r}
dda_maaslin_prevalente <- list()

variables_salida <- c("Survival", "AR", "AHT", "Biliary_complications")

for (i in variables_salida) {

    maslin_objeto <- read_delim(paste0("../../maaslin2/genus/prevalente/paper/", i, "/all_results.tsv")) %>%
    dplyr::filter(value != "RUN 1" , value != "RUN 2")

    dda_maaslin_prevalente[[i]] <- maslin_objeto
    rm(maslin_objeto)

}
```


# Programa pubmed

## Pubmed query

Rescatamos lo que hemos buscado en pubtator

```{r}
pubmed_query <- read.csv("../../Programa genes pubtator/genes_pubmed_genus_paper.csv") |>
    filter(if_all(everything(), ~ . != "")) |>
    drop_na() |>
    dplyr::filter(!grepl("[a-z]", Gene_Name)) |>
    dplyr::filter(!Section %in% c("METHODS", "TABLE"))

```


### Filtrado

```{r, eval=FALSE}
filtered_pubmed_query <- data_frame()
for (i in variables_salida) {
    filtered_genes <- pubmed_query %>%
        dplyr::filter(Variable == i) |>
        group_by(Gene_Name) %>%
        summarise(Unique_PMIDs = n_distinct(PMID)) %>%  # Contar los PMIDs únicos por Gene_Name
        filter(Unique_PMIDs >= 10) %>%  # Quedarte solo con Gene_Name que tienen 10 o más PMIDs
        pull(Gene_Name)  # Extraer la lista de Gene_Name filtrados
    
    if (nrow(filtered_pubmed_query) == 0) {
        # Filtrar el dataframe original usando los genes filtrados
        ft <- pubmed_query %>%
            dplyr::filter(Variable == i) |>
            filter(Gene_Name %in% filtered_genes) 
        
        # Filtro Titulo, abstract, figura
        filtered_pubmed_query <- ft |> 
            dplyr::filter(Section %in% c("TITLE", "ABSTRACT", "FIG")) |>
            distinct()
        
        # Filtro mas de dos veces en Results
        for (j in unique(filtered_pubmed_query$PMID)) {
            genes_resultados <- ft |> 
                dplyr::filter(PMID == j) |> 
                dplyr::filter(Section == "RESULTS") |>
                group_by(Gene_Name) |>
                dplyr::summarise(n = n()) |> 
                dplyr::filter(n >= 2) |> 
                pull(Gene_Name)
            
            filtered_pubmed_query <- ft |> 
                dplyr::filter(PMID == j) |> 
                dplyr::filter(Section == "RESULTS") |>
                dplyr::filter(Gene_Name %in% genes_resultados) |>
                distinct() |> 
                rbind(filtered_pubmed_query)
            
            
        }
        
    } else {
        
        ft <- pubmed_query %>%
            dplyr::filter(Variable == i) |>
            filter(Gene_Name %in% filtered_genes) 
        
        # Filtro Titulo, abstract, figura
        filtered_pubmed_query <- ft |> 
            dplyr::filter(Section %in% c("TITLE", "ABSTRACT", "FIG")) |>
            distinct() |>
            rbind(filtered_pubmed_query)
        
        # Filtro mas de dos veces en Results
        for (j in unique(filtered_pubmed_query$PMID)) {
            genes_resultados <- ft |> 
                dplyr::filter(PMID == j) |> 
                dplyr::filter(Section == "RESULTS") |>
                group_by(Gene_Name) |>
                dplyr::summarise(n = n()) |> 
                dplyr::filter(n >= 2) |> 
                pull(Gene_Name)
            
            filtered_pubmed_query <- ft |> 
                dplyr::filter(PMID == j) |> 
                dplyr::filter(Section == "RESULTS") |>
                dplyr::filter(Gene_Name %in% genes_resultados) |>
                distinct() |> 
                rbind(filtered_pubmed_query)
            
        }
        
    } 
    
}

saveRDS(filtered_pubmed_query, "../../RDSs/filtered_pubmed_query.RDS")
```

Cargamos y mostramos los genes por variable de salida

```{r}
filtered_pubmed_query <- readRDS("../../RDSs/filtered_pubmed_query.RDS")

genes_por_variable <- data_frame()
for (i in variables_salida) {
    top_genes_i <- filtered_pubmed_query |>
            dplyr::filter(Variable == i) |>
            group_by(Gene_ID, Gene_Name) %>%
            summarise(Unique_PMIDs = n_distinct(PMID), .groups = "drop") %>%
            arrange(desc(Unique_PMIDs)) |> 
            dplyr::filter(Unique_PMIDs >= 3)
    
    top_genes_i$Variable <- i  
    
    if (nrow(genes_por_variable) == 0){
        genes_por_variable <- top_genes_i
        
    } else {
        genes_por_variable <- rbind(genes_por_variable, top_genes_i)
        
    }

}

genes_por_variable

```

### Hipo e hiperbiosis

Cargamos y mostramos los genes por variable de salida pero filtrando por hiper e hipo

```{r}
genes_por_variable_biosis <- data_frame()
for (i in variables_salida) {
    taxones_biosis <- dda_maaslin_prevalente[[i]] |> 
        dplyr::filter(pval <= 0.05) |> 
        mutate(biosis = ifelse(coef > 0, "Hiperbiosis", "Hipobiosis")) |>
        dplyr::select(feature, biosis) |>
        dplyr::rename("Taxon" = "feature")
    
    for (j in unique(taxones_biosis$biosis)){
        
         top_genes_i <- filtered_pubmed_query |>
             dplyr::filter(Variable == i) |>
             dplyr::filter(Taxon %in% taxones_biosis$Taxon[taxones_biosis$biosis == j]) |>
             group_by(Gene_Name, Gene_ID) %>%
             summarise(Unique_PMIDs = n_distinct(PMID), .groups = "drop") %>%
             arrange(desc(Unique_PMIDs)) |> 
             dplyr::filter(Unique_PMIDs >= 3)
         
         top_genes_i$Variable <- i  
         top_genes_i$biosis <- j
         
         if (nrow(genes_por_variable_biosis) == 0){
            genes_por_variable_biosis <- top_genes_i
        
        } else {
            genes_por_variable_biosis <- rbind(genes_por_variable_biosis, top_genes_i)
            
        }
        
    }
    
    

}

genes_por_variable_biosis 

```

```{r}
#write.csv(genes_por_variable_biosis, "../../Programa genes pubtator/genes_biosis_para_geneshot_pubmed_query_paper.csv")
```

## Filtrado genes hígado

### GTEx de higado

```{r}
gtex_liver <- read.table("../../../Analisis Lab/Cibersort_DWLS/data_referencias/gene_reads_v10_liver.gct", skip = 2, sep = "\t", header = TRUE, stringsAsFactors = FALSE) 

gtex_liver <- dplyr::rename(gtex_liver, "Gene_name" = "Description")
gtex_liver$Name <- NULL
```

```{r}
mediana_gtex_liver <- gtex_liver %>%
    rowwise() %>%
    mutate(mediana = median(c_across(where(is.numeric)), na.rm = TRUE)) %>%
    ungroup() |> 
    dplyr::select(Gene_name, mediana) |>
    dplyr::filter(mediana > 10)
```

```{r}
ensembl <- useEnsembl(biomart = "genes", 
                      dataset = "hsapiens_gene_ensembl", 
                      version = 110)  
# Obtener los IDs de Ensembl correspondientes
conversion <- getBM(attributes = c("hgnc_symbol", "ensembl_gene_id"),
                    filters = "hgnc_symbol",
                    values = mediana_gtex_liver$Gene_name,
                    mart = ensembl)

names(conversion) <- c("Gene_name", "ENSEMBL")

mediana_gtex_liver <- dplyr::left_join(mediana_gtex_liver, conversion, by = "Gene_name") |>
    mutate(ENSEMBL = if_else(is.na(ENSEMBL), Gene_name, ENSEMBL)) 

```

### Enriquecimiento

<!-- ```{r, eval=FALSE} -->
<!-- enriquecimientos <- list() -->

<!-- variables_salida <- c("Survival", "AR", "AHT", "Biliary_complications") -->

<!-- for (i in variables_salida) { -->

<!--     cat("Starting with: ", i, "\n") -->

<!--     for (j in unique(genes_por_variable_biosis$biosis)){ -->

<!--         lista_genes <- genes_por_variable_biosis |>  -->
<!--             dplyr::filter(Variable == i) |> -->
<!--             dplyr::filter(biosis == j)  -->

<!--         if (nrow(lista_genes) == 0) { next } -->

<!--         genes <- getBM(attributes = c("hgnc_symbol", "ensembl_gene_id", "entrezgene_id"), -->
<!--                        filters = "hgnc_symbol", -->
<!--                        values = lista_genes$Gene_Name, -->
<!--                        mart = ensembl) -->

<!--         genes <- genes |>  -->
<!--             dplyr::filter(ensembl_gene_id %in% mediana_gtex_liver$ENSEMBL)  -->

<!--         cat("\t Starting GO enrichment: ", i, "\n") -->

<!--         ego <- enrichGO(gene          = unique(genes$hgnc_symbol), -->
<!--                             keyType       = "SYMBOL", -->
<!--                             OrgDb         = org.Hs.eg.db, -->
<!--                             ont           = "BP", -->
<!--                             pAdjustMethod = "BH", -->
<!--                             pvalueCutoff  = 0.05, -->
<!--                             qvalueCutoff  = 0.05, -->
<!--                             readable      = TRUE) -->


<!--         cat("\t\t GO ", j, " finished \n") -->

<!--         enriquecimientos[[paste0("GO_",i, "_", j)]] <- ego -->

<!--         cat("\t GO enrichment of ", i, " finished \n") -->

<!--     } -->

<!--     cat("\t\t", i, " finished \n") -->

<!-- } -->


<!-- saveRDS(enriquecimientos, "../../RDSs/enriquecimientos_pubmed_genus.RDS") -->
<!-- ``` -->

```{r}
enrique_pubmed <- readRDS("../../RDSs/enriquecimientos_pubmed_genus.RDS")
```


##### Survival

```{r}
lista_genes <- genes_por_variable_biosis |> 
        dplyr::filter(Variable == "Survival") 
    
genes <- getBM(attributes = c("hgnc_symbol", "ensembl_gene_id", "entrezgene_id"),
               filters = "hgnc_symbol",
               values = lista_genes$Gene_Name,
               mart = ensembl)

genes <- genes |> 
    dplyr::filter(ensembl_gene_id %in% mediana_gtex_liver$ENSEMBL) 

ego <- enrichGO(gene          = unique(genes$ensembl_gene_id),
                keyType       = "ENSEMBL",
                OrgDb         = org.Hs.eg.db,
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05,
                readable      = TRUE)

enrique_pubmed$GO_Survival <- ego
saveRDS(genes, "../../RDSs/genes_higado_busqueda_pubmed_survival.RDS")
```

```{r, fig.width=15, fig.height=15}
enrique_pubmed$GO_Survival@result <- enrique_pubmed$GO_Survival@result %>%
  filter(p.adjust < 0.05) 

edo <- pairwise_termsim(enrique_pubmed$GO_Survival)

treeplot(edo, 
         cluster.params = list(method = "ward.D2",
                               n = 6,
                               label_words_n = 3,
                               label_format = 40),
         geneClusterPanel = "pie", 
         showCategory = 50, 
         color = "qvalue",
         #cex_category = 2,
         fontsize = 5,
         offset.params = list(bar_tree = 50,
                              tiplab = 2,
                              label_words_n = 4,
                              hexpand= 0.2)
         ) + 
    theme(
        legend.position = "none",
        legend.text = element_text(size = 20),
        legend.title = element_blank()
    )

```

```{r}
genes_upregulados_rnaseq <- readRDS("../../RDSs/genes_upregulados_rnaseq.RDS")

enrique_pubmed$GO_Survival@result |>
    # dplyr::select(Description, geneID) |>
    dplyr::filter(grepl(paste(genes_upregulados_rnaseq, collapse = "|"), geneID)) 
```


```{r}
gcSample1 <- list()

for (i in variables_salida){
    for (j in unique(genes_por_variable_biosis$biosis)){
        genes <- genes_por_variable_biosis |>
            dplyr::filter(Variable == i) |>
            dplyr::filter(biosis == j) |>
            pull(Gene_Name)
        
        if (length(genes) == 0){ next }
        
        genes_ensembl_biosis <- getBM(attributes = c("hgnc_symbol", "ensembl_gene_id"),
                              filters = "hgnc_symbol",
                              values = genes,
                              mart = ensembl)
        
        gcSample1[[paste0(i,"_",j)]] <- genes_ensembl_biosis |> 
            dplyr::filter(ensembl_gene_id %in% mediana_gtex_liver$ENSEMBL) |>
            pull(ensembl_gene_id)
            # pull(hgnc_symbol)
    
    }
        
}
```

```{r, fig.width=15, fig.height=20}
ck <- compareCluster(geneCluster = gcSample1[1:2], fun = "enrichGO", keyType = "ENSEMBL", OrgDb = org.Hs.eg.db)
xx <- pairwise_termsim(ck)
id_counts <- xx@compareClusterResult %>%
  group_by(ID) %>%
  summarise(n_clusters = n_distinct(Cluster)) %>%
  filter(n_clusters == 1)

xx@compareClusterResult <- xx@compareClusterResult %>%
    filter(ID %in% id_counts$ID) |>
    filter(!Description %in% c("low-density lipoprotein particle binding"))

p1 <- emapplot(xx, showCategory = 114, 
         pie.params = list(pie = "count"),
         cex.params = list(
             line = 0.2,
             category_label = 1.2,
             category_node = 2
         )
         ) +  
  scale_fill_manual(values = c("#D5695DFF", "#D3BA68FF")) +
    theme(
        legend.position = "top",                   
        legend.title = element_blank(),            
        legend.text = element_text(size = 14, face = "bold"),     
        # legend.key = element_rect(shape = "circle")  
        )


p1
```

```{r}
ck <- compareCluster(geneCluster = gcSample1[1:2], fun = "enrichGO", keyType = "SYMBOL", OrgDb = org.Hs.eg.db)
xx <- pairwise_termsim(ck)
id_counts <- xx@compareClusterResult %>%
  group_by(ID) %>%
  summarise(n_clusters = n_distinct(Cluster)) %>%
  filter(n_clusters == 1)

xx@compareClusterResult <- xx@compareClusterResult %>%
    filter(ID %in% id_counts$ID) |>
    filter(Description %in% c("lipoprotein particle receptor binding", "protein-lipid complex binding", "low-density lipoprotein particle receptor binding", "lipoprotein particle binding", "apolipoprotein binding"))

cnetplot(xx, 
         pie.params = list(pie = "count"),
         node_label = "category",
         node_color = "category",
         layout = "circle",
         cex.params = list(
             category_label = 1.2,
             category_node = 2
             )
         ) +  
    theme(
        legend.position = "none"
    ) 
```

###### Genes upregulados

```{r}
xx <- pairwise_termsim(enrique_pubmed$GO_Survival)

description <- xx@result |>
    dplyr::filter(grepl(paste(genes_upregulados_rnaseq, collapse = "|"), geneID)) |>
    pull(Description)

xx@result 

```


```{r, fig.width=20, fig.height=20}
genes_upregulados_rnaseq <- readRDS("../../RDSs/genes_upregulados_rnaseq.RDS")

ck <- compareCluster(geneCluster = gcSample1[1:2], fun = "enrichGO", keyType = "SYMBOL", OrgDb = org.Hs.eg.db)

xx <- pairwise_termsim(ck)
id_counts <- xx@compareClusterResult %>%
  group_by(ID) %>%
  summarise(n_clusters = n_distinct(Cluster)) %>%
  filter(n_clusters == 1)

xx@compareClusterResult <- xx@compareClusterResult %>%
  filter(ID %in% id_counts$ID)

description <- xx@compareClusterResult |>
    dplyr::filter(grepl(paste(genes_upregulados_rnaseq, collapse = "|"), geneID)) |>
    pull(Description)

cnetplot(xx,
         showCategory = 90, 
         layout = "circle",
         colorEdge = T,
         node_label = "all",
         cex.params = list(
             gene_node = 0.001,
             category_label = 1.5
             ),
         hilight.params = list(
             category = description,
             alpha_no_hilight = 0.1
             )
         ) +
    theme(
        legend.position = "none"
        )

```









```{r, fig.width=10, fig.height=10}
ck <- compareCluster(geneCluster = gcSample1[1:2], fun = "enrichGO", keyType = "SYMBOL", OrgDb = org.Hs.eg.db)
id_counts <- ck@compareClusterResult %>%
    group_by(ID) %>%
    summarise(n_clusters = n_distinct(Cluster)) %>%
    filter(n_clusters == 1)
ck@compareClusterResult <- ck@compareClusterResult %>%
    filter(ID %in% id_counts$ID) |> 
    filter(Cluster == "Survival_Hiperbiosis")

cnetplot(ck, 
         showCategory = 90, 
         layout = "circle",
         colorEdge = T,
         node_label = "all",
         cex.params = list(
             gene_node = 0.001,
             category_label = 1.5
             ),
         hilight.params = list(category = c(
             "heat shock protein binding",
             "Hsp90 protein binding",
             "transcription coactivator binding",
             "histone deacetylase binding",
             "transcription coregulator binding",
             "RNA polymerase II-specific DNA-binding transcription factor binding",
             "DNA-binding transcription factor binding",
             "ubiquitin-like protein ligase binding",
             "ubiquitin protein ligase binding"
             ),
             alpha_no_hilight = 0.1
             )
         ) +
    theme(
        legend.position = "none"
    )


```


```{r, fig.width=10, fig.height=10}
ck <- compareCluster(geneCluster = gcSample1[1:2], fun = "enrichGO", keyType = "SYMBOL", OrgDb = org.Hs.eg.db)
id_counts <- ck@compareClusterResult %>%
    group_by(ID) %>%
    summarise(n_clusters = n_distinct(Cluster)) %>%
    filter(n_clusters == 1)
ck@compareClusterResult <- ck@compareClusterResult %>%
    filter(ID %in% id_counts$ID) |> 
    filter(Cluster == "Survival_Hipobiosis")

cnetplot(ck, 
         showCategory = 94, 
         layout = "circle",
         colorEdge = T,
         node_label = "all",
         cex.params = list(
             gene_node = 0.001,
             category_label = 1.5
             ),
         # hilight.params = list(category = c(
         #     "peptidase inhibitor activity", 
         #     "endopeptidase inhibitor activity",
         #     "endopeptidase regulator activity"
         #     ),
         #     alpha_no_hilight = 0.1
         #     )
         ) +
    theme(
        legend.position = "none"
    )

```




##### AR

```{r}
lista_genes <- genes_por_variable_biosis |> 
        dplyr::filter(Variable == "AR") 
    
    
genes <- getBM(attributes = c("hgnc_symbol", "ensembl_gene_id", "entrezgene_id"),
               filters = "hgnc_symbol",
               values = lista_genes$Gene_Name,
               mart = ensembl)

genes <- genes |> 
    dplyr::filter(ensembl_gene_id %in% mediana_gtex_liver$ENSEMBL) 

ego <- enrichGO(gene          = unique(genes$ensembl_gene_id),
                    keyType       = "ENSEMBL",
                    OrgDb         = org.Hs.eg.db,
                    ont           = "BP",
                    pAdjustMethod = "BH",
                    pvalueCutoff  = 0.05,
                    qvalueCutoff  = 0.05,
                    readable      = TRUE)

enrique_pubmed$GO_AR <- ego
```

```{r, fig.width=15, fig.height=15}
enrique_pubmed$GO_AR@result <- enrique_pubmed$GO_AR@result %>%
  filter(pvalue < 0.05) |>
    dplyr::filter(!Description %in% c("regulation of neuroinflammatory response",
                                      "positive regulation of neuroinflammatory response",
                                      "regulation of branching involved in lung morphogenesis"))


edo <- pairwise_termsim(enrique_pubmed$GO_AR)
# emapplot(edo, layout="kk", showCategory = 50)
treeplot(edo, 
         cluster.params = list(method = "ward.D2",
                               n = 4,
                               label_words_n = 3,
                               label_format = 40),
         geneClusterPanel = "pie", 
         showCategory = 50, 
         color = "qvalue",
         #cex_category = 2,
         fontsize = 5,
         offset.params = list(bar_tree = 50,
                              tiplab = 2,
                              label_words_n = 4,
                              hexpand= 0.2)
         ) + 
    theme(
        legend.position = "none",
        legend.text = element_text(size = 20),
        legend.title = element_blank()
    )

```

```{r, fig.width=15, fig.height=20}
ck <- compareCluster(geneCluster = gcSample1[3:4], fun = "enrichGO", keyType = "ENSEMBL", OrgDb = org.Hs.eg.db)
xx <- pairwise_termsim(ck)
id_counts <- xx@compareClusterResult %>%
  group_by(ID) %>%
  summarise(n_clusters = n_distinct(Cluster)) %>%
  filter(n_clusters == 1)

xx@compareClusterResult <- xx@compareClusterResult %>%
  filter(ID %in% id_counts$ID)

p1 <- emapplot(xx,  
         pie.params = list(pie = "count"),
         cex.params = list(
             line = 0.2,
             category_label = 1.6,
             category_node = 0.3
         )
         ) +  
  scale_fill_manual(values = c("#D5695DFF", "#D3BA68FF")) +
    theme(
        legend.position = "top",                   
        legend.title = element_blank(),            
        legend.text = element_text(size = 14, face = "bold")
        )


p1
```

```{r}
treeplot(xx, geneClusterPanel = "pie", showCategory = 29, hclust_method = "ward.D2", nCluster = 4)
```

##### AHT

```{r}
lista_genes <- genes_por_variable_biosis |> 
        dplyr::filter(Variable == "AHT") 
    
    
genes <- getBM(attributes = c("hgnc_symbol", "ensembl_gene_id", "entrezgene_id"),
               filters = "hgnc_symbol",
               values = lista_genes$Gene_Name,
               mart = ensembl)

genes <- genes |> 
    dplyr::filter(ensembl_gene_id %in% mediana_gtex_liver$ENSEMBL) 

ego <- enrichGO(gene          = unique(genes$ensembl_gene_id),
                    keyType       = "ENSEMBL",
                    OrgDb         = org.Hs.eg.db,
                    ont           = "BP",
                    pAdjustMethod = "BH",
                    pvalueCutoff  = 0.1,
                    qvalueCutoff  = 0.1,
                    readable      = TRUE)

enrique_pubmed$GO_AHT <- ego
```

```{r, fig.width=15, fig.height=15}
enrique_pubmed$GO_AHT@result <- enrique_pubmed$GO_AHT@result %>%
  filter(pvalue < 0.05) 

emapplot(edo)


edo <- pairwise_termsim(enrique_pubmed$GO_AHT)
# emapplot(edo, layout="kk", showCategory = 50)
treeplot(edo, 
         cluster.params = list(method = "ward.D2",
                               n = 5,
                               label_words_n = 3,
                               label_format = 40),
         geneClusterPanel = "pie", 
         showCategory = 50, 
         color = "qvalue",
         #cex_category = 2,
         fontsize = 5,
         offset.params = list(bar_tree = 50,
                              tiplab = 2,
                              label_words_n = 4,
                              hexpand= 0.2)
         ) + 
    theme(
        legend.position = "none",
        legend.text = element_text(size = 20),
        legend.title = element_blank()
    )

```

##### BC

```{r}
lista_genes <- genes_por_variable_biosis |> 
        dplyr::filter(Variable == "Biliary_complications") 
    
    
genes <- getBM(attributes = c("hgnc_symbol", "ensembl_gene_id", "entrezgene_id"),
               filters = "hgnc_symbol",
               values = lista_genes$Gene_Name,
               mart = ensembl)

genes <- genes |> 
    dplyr::filter(ensembl_gene_id %in% mediana_gtex_liver$ENSEMBL) 

ego <- enrichGO(gene          = unique(genes$ensembl_gene_id),
                    keyType       = "ENSEMBL",
                    OrgDb         = org.Hs.eg.db,
                    ont           = "BP",
                    pAdjustMethod = "BH",
                    pvalueCutoff  = 0.05,
                    qvalueCutoff  = 0.05,
                    readable      = TRUE)

enrique_pubmed$GO_BC <- ego
```

```{r, fig.width=15, fig.height=15}
enrique_pubmed$GO_BC@result <- enrique_pubmed$GO_BC@result %>%
  filter(pvalue < 0.05) |>
    dplyr::filter(!Description %in% c("regulation of neuroinflammatory response",
                                      "positive regulation of neuroinflammatory response",
                                      "regulation of branching involved in lung morphogenesis",
                                      "neuroinflammatory response",
                                      "glial cell activation"))


edo <- pairwise_termsim(enrique_pubmed$GO_BC)
# emapplot(edo, layout="kk", showCategory = 50)
treeplot(edo, 
         cluster.params = list(method = "ward.D2",
                               n = 6,
                               label_words_n = 3,
                               label_format = 40),
         geneClusterPanel = "pie", 
         showCategory = 50, 
         color = "qvalue",
         #cex_category = 2,
         fontsize = 5,
         offset.params = list(bar_tree = 50,
                              tiplab = 2,
                              label_words_n = 4,
                              hexpand= 0.2)
         ) + 
    theme(
        legend.position = "none",
        legend.text = element_text(size = 20),
        legend.title = element_blank()
    )

```

```{r, fig.width=15, fig.height=20}
ck <- compareCluster(geneCluster = gcSample1[6:7], fun = "enrichGO", keyType = "ENSEMBL", OrgDb = org.Hs.eg.db)
xx <- pairwise_termsim(ck)
id_counts <- xx@compareClusterResult %>%
  group_by(ID) %>%
  summarise(n_clusters = n_distinct(Cluster)) %>%
  filter(n_clusters == 1)

xx@compareClusterResult <- xx@compareClusterResult %>%
  filter(ID %in% id_counts$ID)

p1 <- emapplot(xx,  
         pie.params = list(pie = "count"),
         cex.params = list(
             line = 0.2,
             category_label = 1.6,
             category_node = 1
         )
         ) +  
  scale_fill_manual(values = c("#D5695DFF", "#D3BA68FF")) +
    theme(
        legend.position = "top",                   
        legend.title = element_blank(),            
        legend.text = element_text(size = 20, face = "bold")
        )


p1

# treeplot(xx, geneClusterPanel = "pie", showCategory = 29)

```

## Coexpression

```{r}

files <- Sys.glob("../../geneshot_predicted_genes_*.tsv")

combined_df <- NULL

for (file in files) {
    nombre <- sub("^.*geneshot_predicted_genes_(.*?)\\.tsv$", "\\1", file)
    parts <- unlist(strsplit(nombre, "_", fixed = TRUE))
    variable <- parts[1]
    biosis <- parts[2]
    
    
    temp_df <- data.table::fread(file, sep = "\t")[,2:4] |>
        rename(
            "publications" = "Publication Count"
        )
    
    temp_df$variable <- ifelse(variable == "BC", "Biliary_complications", variable)
    temp_df$biosis <- biosis
    
    temp_df <- temp_df %>%
        mutate(rareza = case_when(
            publications <= 7 ~ "Rare",
            publications >= 8 & publications <= 25 ~ "Uncommon",
            publications >= 26 & publications <= 87 ~ "Common",
            publications >= 88 ~ "Very Common"
            )
        )
  
    #  Combinar con el dataframe general
    if (is.null(combined_df)) {
        combined_df <- temp_df  
        
    } else {
        combined_df <- bind_rows(combined_df, temp_df)  
    }
}


combined_df

```

```{r, fig.width=15, fig.height=5}
plots <- list()
for (i in variables_salida) {
    for (j in unique(combined_df$biosis)){
        
        if(i == "AHT"){next}
        
        plots[[paste0(i, "_",j)]] <- combined_df %>%
            dplyr::filter(variable == i) |>
            dplyr::filter(biosis == j) |>
            dplyr::count(rareza) |>
            ggplot(aes(x = reorder(rareza, n), y = n, fill = rareza)) +
              geom_bar(stat = "identity") +
              scale_fill_manual(values = c("Rare" = "#1f77b4", "Uncommon" = "#ff7f0e", 
                                           "Common" = "#2ca02c", "Very Common" = "#d62728")) +
              labs(title = i,
                   x = "Rareza",
                   y = "Número de Genes") +
              theme_minimal() +
              theme(legend.position = "none",
                    plot.title = element_text(hjust = 0.5, face = "bold", size = 14),
                    axis.text.x = element_text(face = "bold")) +
            coord_flip()
        
    }
    
    
}

cowplot::plot_grid(plotlist = plots, ncol = 3)
```

#### Enriquecimiento

```{r, eval=FALSE}

enriquecimientos <- list()

variables_salida <- c("Survival", "AR", "AHT", "Biliary_complications")

for (i in variables_salida) {
    if(i == "AHT"){next}
    
    cat("Starting with: ", i, "\n")
    
    for (j in unique(combined_df$biosis)){
    
        lista_genes <- combined_df |> 
            dplyr::filter(variable == i) |>
            dplyr::filter(biosis == j)
        
        lista_enriquecer.df <- bitr(lista_genes$Gene, fromType = "SYMBOL",
                                    toType = c("ENSEMBL", "ENTREZID"),
                                    OrgDb = org.Hs.eg.db)
        
        cat("\t Starting GO enrichment: ", i, "\n")
        
        ego <- enrichGO(gene          = unique(lista_enriquecer.df$SYMBOL),
                            keyType       = "SYMBOL",
                            OrgDb         = org.Hs.eg.db,
                            ont           = "BP",
                            pAdjustMethod = "BH",
                            pvalueCutoff  = 0.1,
                            qvalueCutoff  = 0.1,
                            readable      = TRUE)
        
            
        cat("\t\t GO ", j, " finished \n")
        
        enriquecimientos[[paste0("GO_",i, "_", j)]] <- ego
        
        cat("\t GO enrichment of ", i, " finished \n")
        cat("\t Starting KEGG enrichment: ", i, "\n")
        
        kk <- enrichKEGG(gene         = unique(lista_enriquecer.df$ENTREZID), 
                         organism     = 'hsa',
                         pvalueCutoff = 0.05)
        
        enriquecimientos[[paste0("KEGG_",i, "_", j)]] <- kk
    
    }
    
    cat("\t\t", i, " finished \n")
    
}


saveRDS(enriquecimientos, "../../RDSs/enriquecimientos_pubmed_genus_related_genes.RDS")
```

```{r}
enrique_pubmed_related <- readRDS("../../RDSs/enriquecimientos_pubmed_genus_related_genes.RDS")
```

```{r, fig.width=15, fig.height= 10}
enrique_pubmed_related$GO_Survival_hiperbiosis@result <- enrique_pubmed_related$GO_Survival_hiperbiosis@result |>
    filter(pvalue < 0.05)

ck <- enrique_pubmed_related$GO_Survival_hiperbiosis

xx <- pairwise_termsim(ck)

treeplot(xx, 
         cluster.params = list(
             method = "ward.D2",
             n = 9, 
             label_words_n = 3
             ),
         offset.params = list(
             bar_tree = 50,
             tiplab = 0.5
             ),
         geneClusterPanel = "pie",
         showCategory = 200, 
         color = "pvalue",
         cex_category = 1,
         fontsize = 5
         ) + 
    theme(
        legend.position = "none",
        legend.text = element_text(size = 20),
        legend.title = element_blank()
    )



```

```{r, fig.width=15, fig.height= 10}
enrique_pubmed_related$GO_Survival_hipobiosis@result <- enrique_pubmed_related$GO_Survival_hipobiosis@result |>
    filter(pvalue < 0.05)

ck <- enrique_pubmed_related$GO_Survival_hipobiosis

xx <- pairwise_termsim(ck)

treeplot(xx, 
         cluster.params = list(
             method = "ward.D2",
             n = 8, 
             label_words_n = 3
             ),
         offset.params = list(
             bar_tree = 50,
             tiplab = 0.5
             ),
         geneClusterPanel = "pie",
         showCategory = 200, 
         color = "pvalue",
         cex_category = 1,
         fontsize = 5
         ) + 
    theme(
        legend.position = "none",
        legend.text = element_text(size = 20),
        legend.title = element_blank()
    )



```

```{r, fig.width=10, fig.height=10}
gcSample1 <- list()

for (i in variables_salida){
    if(i == "AHT") {next}
    for (j in unique(combined_df$biosis)){
        gcSample1[[paste0(i,"_",j)]] <- combined_df |>
            dplyr::filter(variable == i) |>
            dplyr::filter(biosis == j) |>
            pull(Gene)
    
    }
        
}
```

```{r, fig.width=10, fig.height=10}

ck <- compareCluster(geneCluster = gcSample1[1:2], fun = "enrichGO", keyType = "SYMBOL", OrgDb = org.Hs.eg.db)
xx <- pairwise_termsim(ck)
id_counts <- xx@compareClusterResult %>%
  group_by(ID) %>%
  summarise(n_clusters = n_distinct(Cluster)) %>%
  filter(n_clusters == 1)

xx@compareClusterResult <- xx@compareClusterResult %>%
  filter(ID %in% id_counts$ID)

p1 <- emapplot(xx, 
         pie.params = list(pie = "count"),
         cex.params = list(
             category_node = 2,
             category_label = 1.5
             )
         ) +  
  scale_fill_manual(values = c("#1f77b4", "#ff7f0e", "#2ca02c", "#d62728", "#9467bd",
                               "#8c564b", "#e377c2"))


dotplot(xx, showCategory = 18)
```

```{r, fig.width=15, fig.height= 10}
treeplot(xx, 
         cluster.params = list(
             method = "ward.D2",
             n = 7, 
             label_words_n = 3
             ),
         offset.params = list(
             bar_tree = 15,
             tiplab = 2
             ),
         geneClusterPanel = "pie",
         showCategory = 94, 
         color = "pvalue",
         cex_category = 10,
         fontsize = 5
         ) + 
    theme(
        legend.position = "top",
        legend.text = element_text(size = 20),
        legend.title = element_blank()
    )

```

```{r, fig.width=500, fig.height=500}
ck <- compareCluster(geneCluster = gcSample1[1:2], fun = "enrichGO", keyType = "SYMBOL", OrgDb = org.Hs.eg.db)
xx <- pairwise_termsim(ck)

cnetplot(xx, 
         showCategory = 94, 
         layout = "circle",
         #colorEdge = T,
         node_label = "all",
         cex.params = list(
             gene_node = 0.1,
             category_node = 0.1,
             category_label = 10,
             gene_label = 20
             ),
         hilight.params = list(category = c(
            "peptidoglycan binding",
            "NADP+ nucleosidase activity","NAD+ nucleotidase, cyclic ADP-ribose generating",
            "tumor necrosis factor receptor superfamily binding",
            "tumor necrosis factor receptor binding",
            "GTP binding", 
            "hydrolase activity, hydrolyzing N-glycosyl compounds", 
            "guanyl nucleotide binding",
            "guanyl ribonucleotide binding",
            "caspase binding",
            "cysteine-type endopeptidase regulator activity involved in apoptotic process"
             ),
             alpha_no_hilight = 0
             )
         ) +
    theme(
        legend.position = "none"
    ) #+ scale_fill_manual(values = c("grey", "grey"))

```


```{r, fig.width=10, fig.height=10}
ck <- compareCluster(geneCluster = gcSample1[1:2], fun = "enrichGO", keyType = "SYMBOL", OrgDb = org.Hs.eg.db)
id_counts <- ck@compareClusterResult %>%
    group_by(ID) %>%
    summarise(n_clusters = n_distinct(Cluster)) %>%
    filter(n_clusters == 1)
ck@compareClusterResult <- ck@compareClusterResult %>%
    filter(ID %in% id_counts$ID) |> 
    filter(Cluster == "Survival_hiperbiosis")

cnetplot(ck, 
         showCategory = 94, 
         layout = "circle",
         colorEdge = T,
         node_label = "all",
         cex.params = list(
             gene_node = 0.001,
             category_label = 1.5
             ),
         hilight.params = list(category = c(
             # "peptidoglycan binding",
             # "NADP+ nucleosidase activity",
             # "NAD+ nucleotidase, cyclic ADP-ribose generating",
             "tumor necrosis factor receptor superfamily binding",
             "tumor necrosis factor receptor binding"
             # "GTP binding", 
             # "hydrolase activity, hydrolyzing N-glycosyl compounds", 
             # "guanyl nucleotide binding",
             # "guanyl ribonucleotide binding",
             # "caspase binding",
             # "cysteine-type endopeptidase regulator activity involved in apoptotic process"
             ),
             alpha_no_hilight = 0.1
             )
         ) +
    theme(
        legend.position = "none"
    )
```


#### GSEA

```{r, eval=FALSE}

enriquecimientos <- list()

variables_salida <- c("Survival", "AR", "AHT", "Biliary_complications")

for (i in variables_salida) {
    if(i == "AHT"){next}
    
    cat("Starting with: ", i, "\n")
    
    for (j in unique(combined_df$biosis)){
        
        lista_genes <- combined_df |> 
            dplyr::filter(variable == i) |> 
            dplyr::filter(biosis == j)
        
        lista_genes <- lista_genes[order(lista_genes$Score, na.last=NA, decreasing = T),]
        
        gene_names <- lista_genes$Gene
        
        entrez_ids <- bitr(gene_names, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")

        gene_list <- setNames(lista_genes$Score, entrez_ids$SYMBOL)
        gene_list <- gene_list[!is.na(names(gene_list))]
        
        gene_list_sorted <- gene_list[order(lista_genes$Score[!is.na(names(gene_list))], decreasing = T)]
        gene_list_sorted <- sort(gene_list_sorted, decreasing = TRUE)  # De mayor a menor importancia
        gsea_result <- gseGO(geneList = gene_list_sorted,
                             keyType = "SYMBOL",
                             ont = "BP", 
                             nPerm = 1000, 
                             minGSSize = 10, 
                             maxGSSize = 500, 
                             pAdjustMethod = "BH", 
                             pvalueCutoff = 1, 
                             verbose = TRUE, 
                             OrgDb = org.Hs.eg.db)
            
        
        cat("\t Starting GO enrichment: ", i, "\n")
        
        ego <- gsea_result <- gseGO(geneList = gene_list_sorted,
                             keyType = "SYMBOL",
                             ont = "BP", 
                             nPerm = 1000, 
                             minGSSize = 10, 
                             maxGSSize = 500, 
                             pAdjustMethod = "BH", 
                             pvalueCutoff = 1, 
                             verbose = TRUE, 
                             OrgDb = org.Hs.eg.db)
        
            
        cat("\t\t GO ", j, " finished \n")
        
        enriquecimientos[[paste0("GO_",i, "_", j)]] <- ego
        
        cat("\t GO enrichment of ", i, " finished \n")
        cat("\t Starting KEGG enrichment: ", i, "\n")

        gene_list <- setNames(lista_genes$Score, entrez_ids$ENTREZID)
        gene_list <- gene_list[!is.na(names(gene_list))]
        
        gene_list_sorted <- gene_list[order(lista_genes$Score[!is.na(names(gene_list))], decreasing = T)]
        gene_list_sorted <- sort(gene_list_sorted, decreasing = TRUE)  # De mayor a menor importancia
        
        kk <- gseKEGG(gene         = gene_list_sorted, 
                      organism     = 'human',
                      minGSSize    = 120,
                      pvalueCutoff = 0.05,
                      verbose      = FALSE)
        
        enriquecimientos[[paste0("KEGG_",i, "_", j)]] <- kk
    
    }
    
    cat("\t\t", i, " finished \n")
    
}


saveRDS(enriquecimientos, "../../RDSs/enriquecimientos_pubmed_genus_related_genes_gsea.RDS")
```

```{r}
enrique_pubmed_related <- readRDS("../../RDSs/enriquecimientos_pubmed_genus_related_genes_gsea.RDS")
enrique_pubmed_related$GO_Survival_hiperbiosis@result <-  enrique_pubmed_related$GO_Survival_hiperbiosis@result |>
    filter(!Description %in% c("central nervous system development", "gliogenesis"))
```

```{r, fig.width=10, fig.height=10}
ero <- enrique_pubmed_related$GO_Survival_hiperbiosis

ero@result <- ero@result %>%
  filter(pvalue < 0.05)

# Crear un dotplot modificando la estética para usar pvalue en lugar de p.adjust
enrichplot::dotplot(ero, color = "pvalue") 
edox <- setReadable(ero, 'org.Hs.eg.db', 'ENTREZID')

#heatplot(ero, showCategory=20)
cnetplot(ero, node_label="all", showCategory = 46, layout = "kk")
#upsetplot(ero)

edo <- pairwise_termsim(ero)
emapplot(edo, layout="fr")
treeplot(edo, hclust_method = "ward.D2", nCluster = 5, showCategory = 46)

gseaplot2(gsea_result, geneSetID = 1:10, title = gsea_result$Description[1])

```

```{r, fig.width=15, fig.height= 10}
ero <- enrique_pubmed_related$GO_Survival_hiperbiosis

ero@result <- ero@result %>%
  filter(pvalue < 0.05)

xx <- pairwise_termsim(ero)

treeplot(xx, 
         cluster.params = list(
             method = "ward.D2",
             n = 6, 
             label_words_n = 3
             ),
         offset.params = list(
             bar_tree = 10,
             tiplab = 0.5
             ),
         geneClusterPanel = "pie",
         showCategory = 200, 
         color = "pvalue",
         cex_category = 1,
         fontsize = 5
         ) + 
    theme(
        legend.position = "none",
        legend.text = element_text(size = 20),
        legend.title = element_blank()
    )

```

```{r, fig.width=20, fig.height=20}
emapplot(xx, 
         showCategory = 94, 
         layout.params = list(
             layout = "kk"
             ),
         node_label = "category",
         cex.params = list(
             category_node = 2, 
             category_label = 1.5,
             line = 0.5
             ),
         edge.params = list(
             show = TRUE, min = 0.2
             ),
         hilight.params = list(category = c(
                # "response to stress",
                # "regulation of transcription by RNA polymerase II",
                # "transcription by RNA polymerase II",
                # "antigen processing and presentation of peptide antigen via MHC class I",
                # "monoatomic cation transport",
                "negative regulation of T cell activation",
                "negative regulation of leukocyte cell-cell adhesion",
                # "positive regulation of hydrolase activity",
                "negative regulation of cell-cell adhesion",
                "negative regulation of immune system process",
                # "response to tumor necrosis factor",
                "negative regulation of lymphocyte activation",
                "negative regulation of multicellular organismal process",
                # "establishment of localization",
                # "response to chemical",
                # "negative regulation of response to biotic stimulus",
                # "localization",
                "negative regulation of leukocyte proliferation",
                # "transport",
                # "positive regulation of transcription by RNA polymerase II",
                "negative regulation of cell adhesion",
                # "response to external stimulus",
                # "mononuclear cell migration",
                # "cellular response to tumor necrosis factor",
                # "interleukin-1 production",
                # "regulation of interleukin-1 production",
                # "calcium-mediated signaling",
                # "defense response",
                # "cellular response to chemical stimulus",
                "negative regulation of leukocyte activation",
                "negative regulation of cell activation",
                # "negative regulation of defense response",
                # "regulation of cell communication",
                # "regulation of cytokine-mediated signaling pathway",
                # "regulation of hydrolase activity",
                # "positive regulation of DNA-templated transcription",
                # "positive regulation of RNA biosynthetic process",
                # "antigen processing and presentation of peptide antigen",
                # "inflammatory response",
                # "regulation of monoatomic ion transport",
                # "regulation of mononuclear cell migration",
                # "antigen processing and presentation of endogenous antigen",
                # "gliogenesis",
                "negative regulation of mononuclear cell proliferation",
                "negative regulation of lymphocyte proliferation"
                # "central nervous system development"
             ),
             alpha_no_hilight = 0.1
             )
         ) +
    theme(
        legend.position = "none"
    )

```

```{r, fig.width=10, fig.height=10}
categories <- c("negative regulation of T cell activation",
                "negative regulation of leukocyte cell-cell adhesion",
                "negative regulation of cell-cell adhesion",
                "negative regulation of immune system process",
                "negative regulation of lymphocyte activation",
                "negative regulation of multicellular organismal process",
                "negative regulation of leukocyte proliferation",
                "negative regulation of cell adhesion",
                "negative regulation of leukocyte activation",
                "negative regulation of cell activation",
                "negative regulation of mononuclear cell proliferation",
                "negative regulation of lymphocyte proliferation")

cnetplot(xx, 
         showCategory = categories, 
         layout = "circle",
         #colorEdge = T,
         node_label = "all",
         cex.params = list(
             gene_node = 0.1,
             category_node = 1,
             category_label = 1,
             gene_label = 1
             )
         ) +
    theme(
        legend.position = "none"
    ) #+ scale_fill_manual(values = c("grey", "grey"))
```

```{r, fig.width=20, fig.height=20}
emapplot(xx, 
         showCategory = 94, 
         layout.params = list(
             layout = "kk"
             ),
         node_label = "category",
         cex.params = list(
             category_node = 2, 
             category_label = 1.5,
             line = 0.5
             ),
         edge.params = list(
             show = TRUE, min = 0.2
             ),
         hilight.params = list(category = c(
                # "response to stress",
                # "regulation of transcription by RNA polymerase II",
                # "transcription by RNA polymerase II",
                # "antigen processing and presentation of peptide antigen via MHC class I",
                # "monoatomic cation transport",
                # "negative regulation of T cell activation",
                # "negative regulation of leukocyte cell-cell adhesion",
                # "positive regulation of hydrolase activity",
                # "negative regulation of cell-cell adhesion",
                # "negative regulation of immune system process",
                "response to tumor necrosis factor",
                # "negative regulation of lymphocyte activation",
                # "negative regulation of multicellular organismal process",
                # "establishment of localization",
                # "response to chemical",
                # "negative regulation of response to biotic stimulus",
                # "localization",
                # "negative regulation of leukocyte proliferation",
                # "transport",
                # "positive regulation of transcription by RNA polymerase II",
                # "negative regulation of cell adhesion",
                # "response to external stimulus",
                # "mononuclear cell migration",
                "cellular response to tumor necrosis factor",
                "interleukin-1 production",
                "regulation of interleukin-1 production",
                # "calcium-mediated signaling",
                # "defense response",
                # "cellular response to chemical stimulus",
                # "negative regulation of leukocyte activation",
                # "negative regulation of cell activation",
                # "negative regulation of defense response",
                # "regulation of cell communication",
                "regulation of cytokine-mediated signaling pathway"
                # "regulation of hydrolase activity",
                # "positive regulation of DNA-templated transcription",
                # "positive regulation of RNA biosynthetic process",
                # "antigen processing and presentation of peptide antigen",
                # "inflammatory response",
                # "regulation of monoatomic ion transport",
                # "regulation of mononuclear cell migration",
                # "antigen processing and presentation of endogenous antigen",
                # "gliogenesis",
                # "negative regulation of mononuclear cell proliferation",
                # "negative regulation of lymphocyte proliferation",
             ),
             alpha_no_hilight = 0.1
             )
         ) +
    theme(
        legend.position = "none"
    )

```

```{r, fig.width=20, fig.height=20}
emapplot(xx, 
         showCategory = 94, 
         layout.params = list(
             layout = "kk"
             ),
         node_label = "category",
         cex.params = list(
             category_node = 2, 
             category_label = 1.5,
             line = 0.5
             ),
         hilight.params = list(category = c(
                "response to stress",
                "regulation of transcription by RNA polymerase II",
                "transcription by RNA polymerase II",
                "antigen processing and presentation of peptide antigen via MHC class I",
                "monoatomic cation transport",
                "negative regulation of T cell activation",
                "negative regulation of leukocyte cell-cell adhesion",
                "positive regulation of hydrolase activity",
                "negative regulation of cell-cell adhesion",
                "negative regulation of immune system process",
                "response to tumor necrosis factor",
                "negative regulation of lymphocyte activation",
                "negative regulation of multicellular organismal process",
                "establishment of localization",
                "response to chemical",
                "negative regulation of response to biotic stimulus",
                "localization",
                "negative regulation of leukocyte proliferation",
                "transport",
                "positive regulation of transcription by RNA polymerase II",
                "negative regulation of cell adhesion",
                "response to external stimulus",
                "mononuclear cell migration",
                "cellular response to tumor necrosis factor",
                "interleukin-1 production",
                "regulation of interleukin-1 production",
                "calcium-mediated signaling",
                "defense response",
                "cellular response to chemical stimulus",
                "negative regulation of leukocyte activation",
                "negative regulation of cell activation",
                "negative regulation of defense response",
                "regulation of cell communication",
                "regulation of cytokine-mediated signaling pathway",
                "regulation of hydrolase activity",
                "positive regulation of DNA-templated transcription",
                "positive regulation of RNA biosynthetic process",
                "antigen processing and presentation of peptide antigen",
                "inflammatory response",
                "regulation of monoatomic ion transport",
                "regulation of mononuclear cell migration",
                "antigen processing and presentation of endogenous antigen",
                "gliogenesis",
                "negative regulation of mononuclear cell proliferation",
                "negative regulation of lymphocyte proliferation",
                "central nervous system development"
             ),
             alpha_no_hilight = 0.1
             )
         ) +
    theme(
        legend.position = "none"
    )

```

```{r}
xx@result |> View()
```


```{r, fig.width=15, fig.height=10}
xx2 <- xx
xx2@result <- xx2@result |>
    filter(Description %in% c(
                "negative regulation of T cell activation",
                "negative regulation of leukocyte cell-cell adhesion",
                "negative regulation of cell-cell adhesion",
                "negative regulation of immune system process",
                "negative regulation of lymphocyte activation",
                "negative regulation of multicellular organismal process",
                "negative regulation of leukocyte proliferation",
                "negative regulation of cell adhesion",
                "negative regulation of leukocyte activation",
                "negative regulation of cell activation",
                "negative regulation of mononuclear cell proliferation",
                "negative regulation of lymphocyte proliferation"
             )
           )
gseaplot2(xx2, 
          geneSetID = 1:12,
          subplots = 1:2,
          color = c("#E495A5", "#86B875", "#7DB0DD"),
          rel_heights = c(1.5, 0.2),
          )

```


```{r, fig.width=15, fig.height= 10}
ero <- enrique_pubmed_related$GO_Survival_hipobiosis

ero@result <- ero@result %>%
  filter(pvalue < 0.05)

xx <- pairwise_termsim(ero)

treeplot(xx, 
         cluster.params = list(
             method = "ward.D2",
             n = 6, 
             label_words_n = 3
             ),
         offset.params = list(
             bar_tree = 10,
             tiplab = 0.5,
             hexpand = 0.4
             ),
         geneClusterPanel = "pie",
         showCategory = 200, 
         color = "pvalue",
         cex_category = 1,
         fontsize = 5
         ) + 
    theme(
        legend.position = "none",
        legend.text = element_text(size = 20),
        legend.title = element_blank()
    )

```

```{r, fig.width=20, fig.height= 20}
emapplot(xx, 
         showCategory = 94, 
         layout.params = list(
             layout = "kk"
             ),
         node_label = "category",
         cex.params = list(
             category_node = 2, 
             category_label = 1.5,
             line = 0.5
             ),
         hilight.params = list(category = c(
                "positive regulation of hydrolase activity",
                "calcium-mediated signaling",
                "exocytosis",
                "response to stress",
                "secretion",
                "regulated exocytosis",
                "response to tumor necrosis factor",
                "regulation of transcription by RNA polymerase II",
                "transcription by RNA polymerase II",
                "cellular response to tumor necrosis factor",
                "antigen receptor-mediated signaling pathway",
                "positive regulation of programmed cell death",
                "neutrophil chemotaxis",
                "regulation of hydrolase activity",
                "positive regulation of catalytic activity",
                "small molecule metabolic process",
                "leukocyte degranulation",
                "localization",
                "establishment of localization",
                "response to interleukin-1",
                "transport",
                "response to external stimulus",
                "regulation of cytokine-mediated signaling pathway",
                "regulation of proteolysis",
                "cellular process",
                "neutrophil migration",
                "positive regulation of type II interferon production",
                "myeloid leukocyte mediated immunity",
                "cellular response to stress",
                "negative regulation of multicellular organismal process",
                "response to biotic stimulus",
                "positive regulation of transcription by RNA polymerase II"
             ),
             alpha_no_hilight = 0.1
             )
         ) +
    theme(
        legend.position = "none"
    )

```

```{r}
xx2@result |> View()
```


```{r, fig.width=15, fig.height=10}
xx2 <- xx
# xx2@result <- xx2@result |>
#     filter(Description %in% c(
#                 "negative regulation of T cell activation",
#                 "negative regulation of leukocyte cell-cell adhesion",
#                 "negative regulation of cell-cell adhesion",
#                 "negative regulation of immune system process",
#                 "negative regulation of lymphocyte activation",
#                 "negative regulation of multicellular organismal process",
#                 "negative regulation of leukocyte proliferation",
#                 "negative regulation of cell adhesion",
#                 "negative regulation of leukocyte activation",
#                 "negative regulation of cell activation",
#                 "negative regulation of mononuclear cell proliferation",
#                 "negative regulation of lymphocyte proliferation"
#              )
#            )
gseaplot2(xx2, 
          geneSetID = 1:12, 
          title = xx$Description[1],
          subplots = 1:2,
          color = c("#E495A5", "#86B875", "#7DB0DD"),
          rel_heights = c(1.5, 0.2),
          )

```