---
title: "OMA metagenomica hígado"
author: "Fernando Lucas Ruiz (fernando.lucas@um.es)"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    highlight: kate
    number_sections: true
    theme: spacelab
    toc: true
    toc_float: true
    code_folding: "hide"
  pdf_document:
    toc: true
subtitle: Cirugía digestiva, endocrina y trasplante de órganos abdominales (IMIB)
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T, warning = F, message = F)
```

```{r}
source("../Librerias.R")
source("../utils.R")
otus_lp_t1_t3 <- readRDS("../../RDSs/paper/otus_lp_t1_t3.RDS")
otus_bilis <- readRDS("../../RDSs/paper/otus_bilis.RDS")
otus_negative_controls <- readRDS("../../RDSs/paper/otus_negative_controls.RDS")
otus_testset <- readRDS("../../RDSs/paper/otus_testset.RDS")

covs <- readRDS("../../RDSs/paper/covs.RDS")
covs_testset <- readRDS("../../RDSs/paper/covs_testset.RDS")
```



# Colapsar taxonomia para random ids

```{r}
taxonomia_combinada_lp_t1_t3 <- otus_lp_t1_t3 %>%
    unite("Combined", Phylum:Species, sep = "_", remove = F)
taxonomia_combinada_bilis <- otus_bilis %>%
    unite("Combined", Phylum:Species, sep = "_", remove = F)
taxonomia_combinada_negative_controls <- otus_negative_controls %>%
    unite("Combined", Phylum:Species, sep = "_", remove = F)
taxonomia_combinada_testset <- otus_testset %>%
    dplyr::filter(codigo != "C-_16s") |>
    unite("Combined", Phylum:Species, sep = "_", remove = F)
taxonomia_combinada_negative_controls_testset <- otus_testset %>%
    dplyr::filter(codigo == "C-_16s") |>
    unite("Combined", Phylum:Species, sep = "_", remove = F)


taxonomia_combinada <- rbind(taxonomia_combinada_lp_t1_t3[,1:6], 
                             taxonomia_combinada_bilis[,1:6], 
                             taxonomia_combinada_negative_controls[,1:6],
                             taxonomia_combinada_testset[,1:6],
                             taxonomia_combinada_negative_controls_testset[,1:6]) 

taxonos <- unique(taxonomia_combinada$Combined)
set.seed(123)  # Para reproducibilidad
random_ids <- replicate(length(taxonos), {
  first_letter <- sample(letters, 1)  # Seleccionar una letra para el primer carácter
  rest <- sample(c(letters, 0:9), 15, replace = TRUE)  # Generar el resto del ID
  paste(c(first_letter, rest), collapse = "")  # Combinar la letra con el resto
})

taxonos_ids <- data.frame(cbind(taxonos, random_ids))

# juntar taxonos_ids a todas_muestras 
otus_lp_t1_t3_combined <- dplyr::left_join(taxonomia_combinada_lp_t1_t3, taxonos_ids, by = c("Combined" = "taxonos"))
otus_bilis_combined <- dplyr::left_join(taxonomia_combinada_bilis, taxonos_ids, by = c("Combined" = "taxonos"))
otus_negative_controls_combined <- dplyr::left_join(taxonomia_combinada_negative_controls, taxonos_ids, by = c("Combined" = "taxonos"))
otus_testset_combined <- dplyr::left_join(taxonomia_combinada_testset, taxonos_ids, by = c("Combined" = "taxonos"))
otus_negative_controls_testset_combined <- dplyr::left_join(taxonomia_combinada_negative_controls_testset, taxonos_ids, by = c("Combined" = "taxonos"))
```

# Controles negativos

```{r}
abundance_table_negative_control <- otus_negative_controls_combined %>%
        dplyr::group_by(random_ids, codigo) %>%
        dplyr::summarize(Count = sum(as.numeric(Count), na.rm = TRUE)) %>%
        ungroup() %>%
        pivot_wider(names_from = codigo, values_from = Count) %>%
        mutate_all(~replace_na(., 0))
    
abundance_table_negative_control <- as.data.frame(abundance_table_negative_control)
rownames(abundance_table_negative_control) <- abundance_table_negative_control$random_ids
abundance_table_negative_control <- abundance_table_negative_control %>% dplyr::select(-random_ids)
```

```{r}
abundance_table_negative_control_testset <- otus_negative_controls_testset_combined %>%
        dplyr::group_by(random_ids, codigo) %>%
        dplyr::summarize(Count = sum(as.numeric(Count), na.rm = TRUE)) %>%
        ungroup() %>%
        pivot_wider(names_from = codigo, values_from = Count) %>%
        mutate_all(~replace_na(., 0))
    
abundance_table_negative_control_testset <- as.data.frame(abundance_table_negative_control_testset)
rownames(abundance_table_negative_control_testset) <- abundance_table_negative_control_testset$random_ids
abundance_table_negative_control_testset <- abundance_table_negative_control_testset %>% dplyr::select(-random_ids)
```

# LP

## Preparar datos

```{r}
samples_LP <- otus_lp_t1_t3_combined %>%
    dplyr::filter(tipo_muestra == "LP") 

pretse_LP <- prepare_TSE_data(samples_LP, covs)
```


## Quitar contaminación

```{r}
otu_combined <- merge(as.data.frame(pretse_LP$abundance_table), as.data.frame(abundance_table_negative_control), by = "row.names", all = TRUE)
rownames(otu_combined) <- otu_combined$Row.names
otu_combined$Row.names <- NULL

# Reemplazar los valores NA por ceros en la matriz combinada
otu_combined[is.na(otu_combined)] <- 0

# Crear un vector que indique qué columnas son controles (TRUE = control, FALSE = muestra real)
is_control <- c(rep(FALSE, ncol(as.data.frame(pretse_LP$abundance_table))), rep(TRUE, ncol(as.data.frame(abundance_table_negative_control))))

# Usar decontam para identificar contaminantes basados en prevalencia
contaminants_prev <- isContaminant(as.matrix(otu_combined), method = "prevalence", neg = is_control) %>%
    filter(contaminant == TRUE & p < 0.05)

head(contaminants_prev)
```

```{r}
if (nrow(contaminants_prev) != 0){
    pretse_LP$abundance_table <- pretse_LP$abundance_table %>% 
        as.data.frame() %>%
        dplyr::select(-rownames(contaminants_prev))
    
    pretse_LP$samples_metadata <- pretse_LP$samples_metadata[rownames(pretse_LP$samples_metadata) != rownames(contaminants_prev),] %>% na.omit()
}
```

## Quitar batch effect

### Preparar datos

```{r}
# taxa tiene que tener la taxonomia en las columnas y los samples en las filas
taxa_LP <- data.frame(t(pretse_LP$abundance_table))

# Para sacar la lista de los batch primero hay que ordenar los samples en el samples_metadata

## Asegurarse de que ambos dataframes tengan los mismos rownames
common_rows <- intersect(rownames(taxa_LP), rownames(pretse_LP$samples_metadata))

taxa_LP <- taxa_LP[common_rows, ]
samples_metadata_LP <- pretse_LP$samples_metadata[match(rownames(taxa_LP), rownames(pretse_LP$samples_metadata)), ]

covar_LP <- samples_metadata_LP 

taxa_LP <- taxa_LP[rownames(taxa_LP) %in% rownames(covar_LP), ]

batchid_LP <- as.factor(covar_LP$Batch)

covar_LP <- covar_LP %>%
    dplyr::select(-c("nombre_muestra", "Batch", "tipo_muestra"))

covar_salida <- covar_LP %>% 
    dplyr::select(MEAF_score, Survival, AR, AHT, Biliary_complications) 
```

### Fine tunning automático

```{r, eval=FALSE}
result_tuned <- Tune_ConQuR(tax_tab=taxa_lp,
                           batchid=batchid_lp,
                           covariates=covar_salida,
                           batch_ref_pool=c("RUN 3"),
                           logistic_lasso_pool=c(T, F),
                           quantile_type_pool=c("standard", "lasso"),
                           simple_match_pool=c(T, F),
                           lambda_quantile_pool=c(NA, "2p/n", "2p/logn"),
                           interplt_pool=c(T, F),
                           frequencyL=0,
                           frequencyU=1,
                           num_core = num_cores-2)
```

```{r}
taxa_corrected_salida <- readRDS("../../RDSs/paper/result_tuned.RDS")
```

Para chequear que todos esta correcto

```{r}
taxa_corrected_salida <- data.frame(t(taxa_corrected_salida$tax_final))
# Quitar la "X" solo de los nombres de columnas que empiezan por "X"
colnames(taxa_corrected_salida) <- sub("^X", "", colnames(taxa_corrected_salida))

rownames(taxa_corrected_salida) <- sub("^X", "", rownames(taxa_corrected_salida))

pretse_LP$abundance_table <- as.matrix(taxa_corrected_salida)

```

## Create a TreeSE

```{r}
# coldata rownames match assay colnames
all(rownames(pretse_LP$samples_metadata) == colnames(pretse_LP$abundance_table)) # our dataset
##  [1] TRUE
class(pretse_LP$samples_metadata) # should be data.frame or DataFrame
##  [1] "data.frame"

# rowdata rownames match assay rownames
all(rownames(pretse_LP$tax) == rownames(pretse_LP$abundance_table)) # our dataset
##  [1] TRUE
class(pretse_LP$tax) # should be data.frame or DataFrame
##  [1] "data.frame"

# Counts
class(pretse_LP$abundance_table) # should be a numeric matrix
##  [1] "matrix" "array"
```

```{r}
# Create a TreeSE
tse_lp <- TreeSummarizedExperiment(
    assays =  SimpleList(counts = pretse_LP$abundance_table),
    colData = DataFrame(pretse_LP$samples_metadata),
    rowData = DataFrame(pretse_LP$tax))
```


# Bilis 

## preparar datos

```{r}
pretse_bilis <- prepare_TSE_data(otus_bilis_combined, covs)
```

## Create tse

```{r}
# coldata rownames match assay colnames
pretse_bilis$samples_metadata <- pretse_bilis$samples_metadata[match(colnames(pretse_bilis$abundance_table), rownames(pretse_bilis$samples_metadata)), ]
all(rownames(pretse_bilis$samples_metadata) == colnames(pretse_bilis$abundance_table)) # our dataset
##  [1] TRUE
class(pretse_bilis$samples_metadata) # should be data.frame or DataFrame
##  [1] "data.frame"

# rowdata rownames match assay rownames
all(rownames(pretse_bilis$tax) == rownames(pretse_bilis$abundance_table)) # our dataset
##  [1] TRUE
class(pretse_bilis$tax) # should be data.frame or DataFrame
##  [1] "data.frame"

# Counts
class(pretse_bilis$abundance_table) # should be a numeric matrix
##  [1] "matrix" "array"
```

```{r}
tse_bilis <- TreeSummarizedExperiment(
    assays =  SimpleList(counts = pretse_bilis$abundance_table),
    colData = DataFrame(pretse_bilis$samples_metadata),
    rowData = DataFrame(pretse_bilis$tax))
```

# Testset 

## preparar datos

```{r}
otus_testset_combined$nombre_muestra <- otus_testset_combined$codigo
otus_testset_combined$tipo_muestra <- "testset"

pretse_testset <- prepare_TSE_data(otus_testset_combined, covs_testset)
```

## Quitar contaminación

```{r}
otu_combined <- merge(as.data.frame(pretse_testset$abundance_table), as.data.frame(abundance_table_negative_control_testset), by = "row.names", all = TRUE)
rownames(otu_combined) <- otu_combined$Row.names
otu_combined$Row.names <- NULL

# Reemplazar los valores NA por ceros en la matriz combinada
otu_combined[is.na(otu_combined)] <- 0

# Crear un vector que indique qué columnas son controles (TRUE = control, FALSE = muestra real)
is_control <- c(rep(FALSE, ncol(as.data.frame(pretse_testset$abundance_table))), rep(TRUE, ncol(as.data.frame(abundance_table_negative_control_testset))))

# Usar decontam para identificar contaminantes basados en prevalencia
contaminants_prev <- isContaminant(as.matrix(otu_combined), method = "prevalence", neg = is_control) %>%
    filter(contaminant == TRUE & p < 0.05)

head(contaminants_prev)
```

## Create tse

```{r}
# coldata rownames match assay colnames
pretse_testset$samples_metadata <- pretse_testset$samples_metadata[match(colnames(pretse_testset$abundance_table), rownames(pretse_testset$samples_metadata)), ]
all(rownames(pretse_testset$samples_metadata) == colnames(pretse_testset$abundance_table)) # our dataset
##  [1] TRUE
class(pretse_testset$samples_metadata) # should be data.frame or DataFrame
##  [1] "data.frame"

# rowdata rownames match assay rownames
all(rownames(pretse_testset$tax) == rownames(pretse_testset$abundance_table)) # our dataset
##  [1] TRUE
class(pretse_testset$tax) # should be data.frame or DataFrame
##  [1] "data.frame"

# Counts
class(pretse_testset$abundance_table) # should be a numeric matrix
##  [1] "matrix" "array"
```

```{r}
tse_testset <- TreeSummarizedExperiment(
    assays =  SimpleList(counts = pretse_testset$abundance_table),
    colData = DataFrame(pretse_testset$samples_metadata),
    rowData = DataFrame(pretse_testset$tax))
```

# Guardar datos

```{r}
saveRDS(tse_lp, "../../RDSs/paper/tse_lp.RDS")
saveRDS(tse_bilis, "../../RDSs/paper/tse_bilis.RDS")
saveRDS(tse_testset, "../../RDSs/paper/tse_testset.RDS")
```




